spectre-project/rusty-spectre
# [RK] Batch upstream merge

This pull request merges updates from the [Rusty-Kaspa](https://github.com/kaspanet/rusty-kaspa) repository into [Rusty-Spectre](https://github.com/spectre-project/rusty-spectre). It includes several improvements to the mempool, updates for TN-11, as well as enhancements to PSKT/PSST and WRPCv2, as we discussed.

### Important Notes and To-Dos:
- currently build fails due to `openssl`. This will be resolved in a later pull request, which will replace `openssl` with `rusttls`
- lints is failing because of unused code in:
  - `consensus/core/src/tx/script_public_key.rs`
  - `crypto/addresses/src/lib.rs`
  
While this can be bypassed this with `#[allow(dead_code)]`, we should investigate why this code is unused and if it's needed.
I have already addressed the Rust 1.81 lints in 052e65cea752d7e7d1bb9d859a62521b1070fbd6

* `get_server_info()`:
  * broken due to data structure changes

* `wRPC v2 Breaking Changes`:
  * new public node infrastructure deployment
  * custom serializers for elastic data structure serialization
  
- enter our own scripted transactions details in (we can already take some details from `wallet\psst\examples\multisig.rs` and https://explorer.spectre-network.org/txs/76ab09236a7aa13fe8f5389c74545c0110d4bba1ca778e49367d4d8de827834b):
  - `consensus/src/processes/transaction_validator/transaction_validator_populated.rs`
  - `crypto/txscript/src/standard/multisig.rs`
  - `wallet/psst/src/bundle.rs`
  
- fixed readme layout
##

https://github.com/kaspanet/rusty-kaspa/pull/499
* replace by fee on mempool 1b0956d04607b7d769c17bf4f55d3f11433c42de

https://github.com/kaspanet/rusty-kaspa/pull/479
* input signature_script checking to submitTransaction 817a3db7b89c7f364298952e73dea4bec8c0b13f

https://github.com/kaspanet/rusty-kaspa/pull/512
* hint message for P2P reject reason block not found a7ff129bbc6b4c0c2c8d2800f9a2ca869ce34aa6

https://github.com/kaspanet/rusty-kaspa/pull/518
* lazy load origin children 91f81179de9c1fe81ac38d0e752d4b1ce2a5106b

https://github.com/kaspanet/rusty-kaspa/pull/513
* O(k log n) mempool transaction sampler + Fee estimation API 93fdc316f5bd71ff3759c37927815da3299d62fb

https://github.com/kaspanet/rusty-kaspa/pull/519
* update mimalloc dep ec3b0645c8d6fae4fb6eb1a1f4bac12b4b47dd97
  * fixes compiling on debian

https://github.com/kaspanet/rusty-kaspa/pull/523
* feat: next_block_template_feerate f24bbb7f7cf81fd1ca1cf7fe801df32170585a32

https://github.com/kaspanet/rusty-kaspa/pull/526
* semaphore tracing feature (for tracing prune readers vs writers time) 411bc42aa7b3da8f8f88316f305872442c9ed1eb

https://github.com/kaspanet/rusty-kaspa/pull/476
* TN11 bug fix: activate mass hashing when modifying a cached block template 6b46dc664dcc824e538b8798e3a164a863d0bf42
  * fix transaction hash merkle root calculation in modify block template to consider storage mass activation
  * avoid similar future errors: expose only a single calc_hash_merkle_root function with `include_mass_field` arg and update all test usages explicitly
  * move subnet checks to inner location
  * reorganize cache mem estimators

https://github.com/kaspanet/rusty-kaspa/pull/527
* optimizations related to multi-level relations 694da52b710c13c9235fcfdc933b99b048d06317
  * In `ParentsBuilder` optimize the execution path of the common case for high levels
  * Prune relations for all levels below the affiliated proof level for block
  * Keep relations for multi-level parents of pruning point and its anticone

https://github.com/kaspanet/rusty-kaspa/pull/530
* query all DNS seeders if missing many connections 4e35e0bfd22256aa756075876d0c6705039ed702

https://github.com/kaspanet/rusty-kaspa/pull/506
* wRPC v2, integrations, typescript-v2, resolver v2 052e65cea752d7e7d1bb9d859a62521b1070fbd6
  * refactored Borsh serialization (wRPC v2)
  * refactored watch-only
  * fix Rust v1.81 lints
  * using latest resolver APIs (guards)
  * updated TypeScript bindings (typescript-v2)
  * fixes related to "insufficient-funds"
  * `Hex View for ScriptBuilder`:
  * formatted hex view for scripts

* `Kip9 Updates`:
  * improvements to mass calculator (Kip9::Beta variant)

New RPC Methods
* `get_sync_status()`:
  * node sync status polling
* `get_connection_count()`:
  * total TCP sockets in use
* `get_system_info()`:
  * unique ID, git commit hash, total RAM, CPU cores, internal FD limit
* `get_metrics()`:
  * provisional dictionary*based map for custom metrics

https://github.com/kaspanet/rusty-kaspa/pull/528
* get_current_block_color RPC utility 8d49801503fb47dc1d1f8f73e4b075fb549bc4e6
  * getCurrentBlockColor algorithm and RPC functions
  * move `get_current_block_color` to consensus and apply standard sorting
  * check for block existence on `get_current_block_color`
  * add extra DAG order checks to ensure about children
  * Includes:
    1. Stylistic changes using ?
    2. `is_dag_ancestor_of(a, b)` is different than `!is_dag_ancestor_of(b, a)`
  * Bug fix:
    1. Hash -> child
    2. Make store calls only where they are actually used (within the if)
  * Style:
    1. Use struct unfolding syntax
    2. Use a name such as `decedent` which reflects the relation to `hash`
  * remove Borsh derivations
  * fix gRPC message codes
  * fix gRPC `getCurrentBlockColorResponse`
  * improve tests

https://github.com/kaspanet/rusty-kaspa/pull/524
* mempool tweaks 1ced37a318c09cb1e35d672c2440916a3b2a5fa9
  * add case to test_evict
  * remove explicit check for too big transactions
  * raise DEFAULT_MEMPOOL_SIZE_LIMIT to 1GB
  * move to estimated size
  * remove estimated_size field

https://github.com/kaspanet/rusty-kaspa/pull/535
* transaction pool -- fix rare cases where byte size tracking was inaccurate 509e900b04f3e87ce8e209b8ad7faf427552a04f
  * remove crate-level mutable access to inner mempool tx
  * fix bytes size tracking
  * update `update_revalidated_transaction` only if validation result is OK
  * remove transaction if validation fails

https://github.com/kaspanet/rusty-kaspa/pull/537
* feerate estimator outliers: capture block slot by the specific mass of the outlier transaction 7ecb3f40c3a8d7a2cab68656ac89cbae8ba471cb
  * deals with outliers (high feerate transactions)
  * captures block space
  * estimates feerate for remaining block space
  * inaccurate average tx mass for outliers
  * fix: Capture block space by specific mass of outlier transaction
  * includes test to verify changes

https://github.com/kaspanet/rusty-kaspa/pull/538
* refactoring of mass calculator utilities c759870a76ae7573d8abe4378b1b26c0c5b83481
  * rename verbose data mass to `compute_mass`
  * fix TS interface verbose transaction data property name
  * relocate consensus core processes `MassCalculator` to consensus core
  * WASM fix ECDSA pubkey generation
  * error variant related to ECDSA from `PublicKey` creation
  * refactor client mass APIs
    * propagate `minimum_signatures` via pending tx
    * rename mass calc functions to `unsigned` for clarity
  * introduce change index to PTX
  * fix missing transfer of client Transaction mass to `RpcTransaction`
  * introduce `IRawBlock` and `IRawHeader` TS types used in `GetBlockTemplateResponse` and `SubmitBlockRequest`
  * fix docs
  * refactor mass calculator and extract global storage mass function that operates on iterators + use from wallet mass calculator
